Fix spelling and grammatical errors. Suggest changes to some explanat… (#9704)
TheOdinProject/curriculum: new_course/1_programming_fundementals/methods.md

### Ruby's Built-in Methods

> One of Ruby's great advantages for new programmers is it's large number of built-in methods. You've been using many of them already, probably without even realizing it. Over the course of your learning so far, you have modified strings and other objects in various ways. For example, the [`#times`](https://ruby-doc.org/core-2.6.1/Integer.html#method-i-times) and [`#upto`](https://ruby-doc.org/core-2.6.1/Integer.html#method-i-upto) loops that you learned about in the Loops lesson are both methods that are included as part of Ruby's [`Integer`](https://ruby-doc.org/core-2.6.1/Integer.html) class.

Author: Grammatical error: _it's_ should be _its_ as shown in the original code.

One of Ruby's great advantages for new programmers is its large number of built-in methods. You've been using many of them already, probably without even realizing it. Over the course of your learning so far, you have modified strings and other objects in various ways. For example, the [`#times`](https://ruby-doc.org/core-2.6.1/Integer.html#method-i-times) and [`#upto`](https://ruby-doc.org/core-2.6.1/Integer.html#method-i-upto) loops that you learned about in the Loops lesson are both methods that are included as part of Ruby's [`Integer`](https://ruby-doc.org/core-2.6.1/Integer.html) class.

> By the way, if you're wondering why we add a `#` in front of method names, they're just [the convention](https://stackoverflow.com/questions/736120/why-are-methods-in-ruby-documentation-preceded-by-a-hash-sign) for writing out Ruby instance methods. We can use it to write out the full name of an instance method, e.g., `Integer#upto`, or just the method name, e.g., `#upto`, depending on the context.

Reviewer: This is a tonal change, which is fine, but if we're going to go straight, I'd rather go all the way. I'd also prefer to not say "why we add..." since we don't actually have an explanation for "why" other than "it's convention". 

```suggestion
If you're wondering about all of the pound signs (`#`), they're just [the convention](https://stackoverflow.com/questions/736120/why-are-methods-in-ruby-documentation-preceded-by-a-hash-sign) for writing out Ruby instance methods. We can use them to write out the full name of an instance method, e.g., `Integer#upto`, or just the method name, e.g., `#upto`, depending on the context. Note that in the development world, you shouldn't call these ["hashtags"](https://help.twitter.com/en/using-twitter/how-to-use-hashtags). If you want to be super awesome, though, you can call them ["octothorpes"](https://en.wiktionary.org/wiki/octothorpe). That word is totally trending.
```

Here are some things you are not allowed to do with your method names:

* You cannot name your method one of Ruby's approximately 40 reserved words, such as `end`, `while`, or `for`. Checkout the full list [here](http://www.java2s.com/Code/Ruby/Language-Basics/Rubysreservedwords.htm).
* You cannot use any symbols other than `_`, `?`, `!`, and `=`.
* You cannot use `?`, `!`, or `=` anywhere other than at the end of the name.
* You cannot begin a method name with a number.

Here are some examples of valid and invalid method names:

> ~~~

Reviewer 2: We need ruby after the `~~~` for our markdown parse on the main site to be able to format code blocks correctly. Can you add that back in here please? 😄

Author: Interesting. I had taken out `ruby` because it was syntax highlighting the names, which I thought was a bit confusing (names appeared in blue, black, red or a combination of colors). But I can definitely put it back as soon as I get back on my computer.

Reviewer 2: ah I see, in that case I'd leave it as is. It will just format as standard uniform code. This was a good change. Thanks @leonardovega

Author: Thanks @KevinMulhern for guiding me through this process. I know you and the rest of the TOP team are very busy. So I appreciate you guys making the contribution process as welcoming as possible.

method_name      # valid
_name_of_method  # valid
1_method_name    # invalid (Starts with a number)
method_27        # valid

> method?_name     # invalid (Valid symbols. such as ?, can only be used at the end of the name)

Author: Should read:

method?_name     # invalid (Valid symbol, such as "?", can only be used at the end of the name)

puts my_name #=> "Joe Smith"
~~~

This can be useful for example in a situation where you want your method to check for possible errors before continuing:

~~~ruby
def even_odd(number)
  unless number.is_a? Numeric
    return "A number was not entered"
  end

  if number % 2 == 0
    "That is an even number."
  else
    "That is an odd number."
  end
end

puts even_odd(20) #=>  That is an even number.
puts even_odd("Ruby") #=>  A number was not entered.
~~~

Try removing the explicit `return` from the method above. Did the method run as expected?

#### Difference Between `puts` and `return`

A common source of confusion for new programmers is the difference between `puts` and `return`.

 * `puts` is a method that simply prints a given value to the console.

>  * `return` produces a value that can be used in your code. 

Reviewer: I prefer the original versions here since they're a bit more explicit in their explanations.

> For example:

Reviewer: ```suggestion
For example, we can write a method that calculates the square of a number and then `puts` the output to the console.
```

~~~ruby

> def squared(number)

Reviewer: ```suggestion
def puts_squared(number)
```

  puts number * number
end
~~~

> The above method will only print to the console the value of a given number squared. However, it will not produce or return that value.  So writing something like `x = squared(20)` will print `400` to the console but the variable `x` will not be assigned the value `400` but rather it will have a `nil` value.  This is because the method did not actually produce a value.

Reviewer: ```suggestion
This method only prints the value that it calculated to the console, but it doesn't return that value. If we then write `x = squared(20)`, the method will print `400` in the console, but the variable `x` will be assigned a value of `nil`. (If you need a refresher on this, go back and review the Ruby Input and Output lesson.)
```

> If we instead write the same method using a implicit or explicit return:

Reviewer: ```suggestion
Now, let's write the same method but with an implicit return instead of `puts`. (Using an explicit return will act exactly the same in this example.)
```

~~~ruby

> def squared(number)

Reviewer: ```suggestion
def return_squared(number)
```

Reviewer: I gave these two methods different names for clarity.

Author: It's funny because I had originally named those methods "puts_squared" and "return_squared" :blush:

  number * number
end
~~~

> You can then interact with the returned value in your code:

Reviewer: ```suggestion
When we run the `return_squared` method, it won't print any output to the console. Instead, it will `return` the result in a way that allows it to be used in the rest of your code. We can save the output of running this method in a variable (`x` in the code below) and use that variable in a variety of ways. If we would still like to see the result of the method in the console, we can `puts` that variable to the console using string interpolation. 
```

~~~ruby

> x = squared(20) #=> 400

Reviewer: ```suggestion
x = return_squared(20) #=> 400
```

When we call a method on an object, such as our `whisper` string above, it does not modify the original value of that object. A general rule in programming is that you do not want your methods to overwrite the objects that you call them on. This protects you from irreversibly overwriting your data by accident. You *are* able to overwrite your data by explicitly re-assigning a variable (such as `whisper = whisper.downcase`). Another way to do this type of reassignment is with **bang methods**, which are denoted with an exclamation mark (`!`) at the end of the method name.

> By adding a `!` to the end of your method, you indicate that this method should perform its action and simultaneously overwrite the value of the original object with the result.

Reviewer: You're right that there was a typo in the original, but I'd rather go the other way to solve it:

```suggestion
By adding a `!` to the end of your method, you indicate that this method performs its action and simultaneously overwrites the value of the original object with the result.
```

2. To get a different take, read the part of the Objects section that discusses [Methods](http://ruby-for-beginners.rubymonstas.org/objects/methods.html) from Ruby Monsta's *Ruby for Beginners*.
3. For more depth on how you can write your own methods, read the section on [Writing Methods](http://ruby-for-beginners.rubymonstas.org/writing_methods.html) from Ruby Monsta's *Ruby for Beginners*. 
</div> 

### Additional Resources
This section contains helpful links to other content. It isn't required, so consider it supplemental for if you need to dive deeper into something.

  - For a deeper look at methods, read the [Methods chapter](http://ruby.bastardsbook.com/chapters/methods/) from the Bastards Book of Ruby. Try to complete the exercises throughout the chapter.
  - For more discussion on the subtle differences between methods and functions and how they can differ between programming languages, here is a [handy explanation on Stack Overflow](https://stackoverflow.com/questions/155609/difference-between-a-method-and-a-function).

### Knowledge Check

> This section contains questions you can use to check your understanding of this lesson. If you're having trouble answering the questions below on your own, review the material above to find the answer.

Reviewer: The language in the original makes reviewing these questions sound like more of a requirement and less of a suggestion, which I feel is preferable. It's also a change that we would have to make across many lessons, so I'm not sure it's worth it. >.<
